### ISSUE TITLE
预测代码

### ISSUE THREAD

**Issue by HongTu0319** [liucongg/ChatGLM-Finetuning#111]
请问一下作者，有预测的代码吗，可以上传一下不

**Comment by Rorschach2333** [comment]
可以参考GLM2官方给出的方法，我这里结合两边，把调用权重的部分拿出来了，代码全部来自GLM2官方，这里供大家参考，我自己也记录一下。按照刘大的设置来微调的话，lora留下的就是AB升降维矩阵的参数，这里直接调用官方的PeftModel方法合并原模型和lora权重即可；
PT的方法稍微麻烦一点，pt留下的两个权重一大一小我估计分别是原权重和更新后的encoder-prefix的权重，可以在权重索引里面看到；要做的就是把GLM原权重库里的几个构成模型的py文件复制到微调后的权重里，然后tokenizer继续使用官方的，我不知道这样会不会有问题，我做出来效果比较差，但是能跑。

**Comment by Rorschach2333** [comment]
from transformers import AutoModel, AutoTokenizer
import streamlit as st
from peft import PeftModel

st.set_page_config(
    page_title="ChatGLM2-6b 演示",
    page_icon=":robot:",
    layout='wide'
)

#streamlit run web2.py

@st.cache_resource
def get_model():
    #origin
    # tokenizer = AutoTokenizer.from_pretrained("/media/wsq/wsq_project/GLM2/GLM2_model/chatglm2-6b",local_files_only=True,trust_remote_code=True)
    # model = AutoModel.from_pretrained("/media/wsq/wsq_project/GLM2/GLM2_model/chatglm2-6b", local_files_only=True,trust_remote_code=True).cuda()
    #lora hugging官方给好了lora微调的库，训练完成后调用PeftModel即可
    tokenizer = AutoTokenizer.from_pretrained("/media/wsq/wsq_project/GLM2/GLM2_model/chatglm2-6b",local_files_only=True,trust_remote_code=True)
    model = AutoModel.from_pretrained("/media/wsq/wsq_project/GLM2/GLM2_model/chatglm2-6b", local_files_only=True,trust_remote_code=True).cuda()
    model = PeftModel.from_pretrained(model, "/media/wsq/wsq_project/GLM2/ChatGLM-Finetuning-master/output-glm2-lora/epoch-2-step-720")
    
    #PT pt的权重分为一大一小两个文件，大的相当于原权重，小的是修改后的encoder-prefix
    #实测pt后模型的遗忘情况很严重，很难进行正常对话，对prompt要求更高，lora没有这个问题
    # tokenizer = AutoTokenizer.from_pretrained("/media/wsq/wsq_project/GLM2/GLM2_model/chatglm2-6b",local_files_only=True,trust_remote_code=True)
    # model = AutoModel.from_pretrained("/media/wsq/wsq_project/GLM2/ChatGLM-Finetuning-master/output-glm2-PT/epoch-2-step-720", local_files_only=True,trust_remote_code=True).cuda()
    
    # 多显卡支持，使用下面两行代替上面一行，将num_gpus改为你实际的显卡数量
    # from utils import load_model_on_gpus
    # model = load_model_on_gpus("THUDM/chatglm2-6b", num_gpus=2)
    model = model.eval()
    return tokenizer, model


tokenizer, model = get_model()

st.title("ChatGLM2-6B")

max_length = st.sidebar.slider(
    'max_length', 0, 32768, 8192, step=1
)
top_p = st.sidebar.slider(
    'top_p', 0.0, 1.0, 0.8, step=0.01
)
temperature = st.sidebar.slider(
    'temperature', 0.0, 1.0, 0.8, step=0.01
)

if 'history' not in st.session_state:
    st.session_state.history = []

if 'past_key_values' not in st.session_state:
    st.session_state.past_key_values = None

for i, (query, response) in enumerate(st.session_state.history):
    with st.chat_message(name="user", avatar="user"):
        st.markdown(query)
    with st.chat_message(name="assistant", avatar="assistant"):
        st.markdown(response)
with st.chat_message(name="user", avatar="user"):
    input_placeholder = st.empty()
with st.chat_message(name="assistant", avatar="assistant"):
    message_placeholder = st.empty()

prompt_text = st.text_area(label="用户命令输入",
                           height=100,
                           placeholder="请在这儿输入您的命令")

button = st.button("发送", key="predict")

if button:
    input_placeholder.markdown(prompt_text)
    history, past_key_values = st.session_state.history, st.session_state.past_key_values
    for response, history, past_key_values in model.stream_chat(tokenizer, prompt_text, history,
                                                                past_key_values=past_key_values,
                                                                max_length=max_length, top_p=top_p,
                                                                temperature=temperature,
                                                                return_past_key_values=True):
        message_placeholder.markdown(response)

    st.session_state.history = history
    st.session_state.past_key_values = past_key_values

**Comment by Rorschach2333** [comment]
```
from transformers import AutoModel, AutoTokenizer
import streamlit as st
from peft import PeftModel

st.set_page_config(
    page_title="ChatGLM2-6b 演示",
    page_icon=":robot:",
    layout='wide'
)

#streamlit run web2.py

@st.cache_resource
def get_model():
    #origin
    # tokenizer = AutoTokenizer.from_pretrained("/media/wsq/wsq_project/GLM2/GLM2_model/chatglm2-6b",local_files_only=True,trust_remote_code=True)
    # model = AutoModel.from_pretrained("/media/wsq/wsq_project/GLM2/GLM2_model/chatglm2-6b", local_files_only=True,trust_remote_code=True).cuda()
    #lora hugging官方给好了lora微调的库，训练完成后调用PeftModel即可
    tokenizer = AutoTokenizer.from_pretrained("/media/wsq/wsq_project/GLM2/GLM2_model/chatglm2-6b",local_files_only=True,trust_remote_code=True)
    model = AutoModel.from_pretrained("/media/wsq/wsq_project/GLM2/GLM2_model/chatglm2-6b", local_files_only=True,trust_remote_code=True).cuda()
    model = PeftModel.from_pretrained(model, "/media/wsq/wsq_project/GLM2/ChatGLM-Finetuning-master/output-glm2-lora/epoch-2-step-720")
    
    #PT pt的权重分为一大一小两个文件，大的相当于原权重，小的是修改后的encoder-prefix
    #实测pt后模型的遗忘情况很严重，很难进行正常对话，对prompt要求更高，lora没有这个问题
    # tokenizer = AutoTokenizer.from_pretrained("/media/wsq/wsq_project/GLM2/GLM2_model/chatglm2-6b",local_files_only=True,trust_remote_code=True)
    # model = AutoModel.from_pretrained("/media/wsq/wsq_project/GLM2/ChatGLM-Finetuning-master/output-glm2-PT/epoch-2-step-720", local_files_only=True,trust_remote_code=True).cuda()
    
    # 多显卡支持，使用下面两行代替上面一行，将num_gpus改为你实际的显卡数量
    # from utils import load_model_on_gpus
    # model = load_model_on_gpus("THUDM/chatglm2-6b", num_gpus=2)
    model = model.eval()
    return tokenizer, model


tokenizer, model = get_model()

st.title("ChatGLM2-6B")

max_length = st.sidebar.slider(
    'max_length', 0, 32768, 8192, step=1
)
top_p = st.sidebar.slider(
    'top_p', 0.0, 1.0, 0.8, step=0.01
)
temperature = st.sidebar.slider(
    'temperature', 0.0, 1.0, 0.8, step=0.01
)

if 'history' not in st.session_state:
    st.session_state.history = []

if 'past_key_values' not in st.session_state:
    st.session_state.past_key_values = None

for i, (query, response) in enumerate(st.session_state.history):
    with st.chat_message(name="user", avatar="user"):
        st.markdown(query)
    with st.chat_message(name="assistant", avatar="assistant"):
        st.markdown(response)
with st.chat_message(name="user", avatar="user"):
    input_placeholder = st.empty()
with st.chat_message(name="assistant", avatar="assistant"):
    message_placeholder = st.empty()

prompt_text = st.text_area(label="用户命令输入",
                           height=100,
                           placeholder="请在这儿输入您的命令")

button = st.button("发送", key="predict")

if button:
    input_placeholder.markdown(prompt_text)
    history, past_key_values = st.session_state.history, st.session_state.past_key_values
    for response, history, past_key_values in model.stream_chat(tokenizer, prompt_text, history,
                                                                past_key_values=past_key_values,
                                                                max_length=max_length, top_p=top_p,
                                                                temperature=temperature,
                                                                return_past_key_values=True):
        message_placeholder.markdown(response)

    st.session_state.history = history
    st.session_state.past_key_values = past_key_values
```

**Comment by Rorschach2333** [comment]
才发现pull区已经有大佬贴了评估的代码了。。。我这个可以测测遗忘现象吧哈哈

**Comment by liucongg** [comment]
已更新
